spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraRDDPartitionGenerator.scala

```diff
@@ -0,0 +1,323 @@
+package com.datastax.spark.connector.rdd.partitioner
+
+import java.net.InetAddress
+import java.util.logging.Logger
+
+import scala.collection.JavaConversions._
+import scala.collection.parallel.ForkJoinTaskSupport
+import scala.concurrent.forkjoin.ForkJoinPool
+import org.apache.spark.{Logging, Partitioner}
+import com.datastax.driver.core.{Metadata, TokenRange => DriverTokenRange}
+import com.datastax.spark.connector.PartitionKeyColumns
+import com.datastax.spark.connector.cql.{CassandraConnector, TableDef}
+import com.datastax.spark.connector.rdd.partitioner.dht.{Token, TokenFactory}
+import com.datastax.spark.connector.util.Quote._
+import com.datastax.spark.connector.writer.RowWriterFactory
+
+import scala.reflect.ClassTag
+import scala.util.Try
+
+/**
+  * A [[org.apache.spark.Partitioner]] implementation which performs the inverse
+  * operation of a traditional C* hashing. Requires the Key type and the token
+  * value type V.
+  *
+  * Will take objects of type Key and determine given the token ragnes in `indexedRanges`
+  * which range the Key would belong in given the C* schema in `TableDef`
+  *
+  * Under the hood uses a bound statement to generate routing keys which are then
+  * used the driver's internal token factories to determine the token for the
+  * routing key.
+  */
+private[connector] class CassandraRDDPartitioner[Key : ClassTag, V](
+  inputPartitions: Seq[CassandraPartition],
+  val indexedRanges: Seq[(Int, Seq[(Token[V], Token[V], Boolean)])],
```

Comment:
Yeah i mentioned this above with Jacek. The dependency between ranges are neccessary but not sufficent to build CassandraPartitions, CassandraPartitions lack the type information to recreate Ranges. I'll work on making a better super class  for Ranges that can go to CassandraPartitions